src/dh/dhe.ts

```diff
@@ -0,0 +1,235 @@
+import type { dh as DhcType } from '@deephaven/jsapi-types';
+import type {
+  EnterpriseDhType as DheType,
+  EditableQueryInfo,
+  EnterpriseClient,
+  QueryInfo,
+} from '@deephaven-enterprise/jsapi-types';
+import { DraftQuery, QueryScheduler } from '@deephaven-enterprise/query-utils';
+import type {
+  ConsoleType,
+  IdeURL,
+  QuerySerial,
+  UniqueID,
+  WorkerInfo,
+  WorkerURL,
+} from '../types';
+
+const INTERACTIVE_CONSOLE_QUERY_TYPE = 'InteractiveConsole';
+
+export type IDraftQuery = EditableQueryInfo & {
+  isClientSide: boolean;
+  draftOwner: string;
+};
+
+/**
+ * Create DHE client.
+ * @param dhe DHE JsApi
+ * @param serverUrl Server URL
+ * @returns A promise that resolves to the DHE client.
+ */
+export async function createDheClient(
+  dhe: DheType,
+  serverUrl: URL
+): Promise<EnterpriseClient> {
+  const dheClient = new dhe.Client(serverUrl.toString());
+
+  return new Promise(resolve => {
+    const unsubscribe = dheClient.addEventListener(
+      dhe.Client.EVENT_CONNECT,
+      () => {
+        unsubscribe();
+        resolve(dheClient);
+      }
+    );
+  });
+}
+
+/**
+ * Get credentials for a Core+ worker associated with a given DHE client.
+ * @param client The DHE client.
+ * @returns A promise that resolves to the worker credentials.
+ */
+export async function getWorkerCredentials(
+  client: EnterpriseClient
+): Promise<DhcType.LoginCredentials> {
+  const token = await client.createAuthToken('RemoteQueryProcessor');
+  return {
+    type: 'io.deephaven.proto.auth.Token',
+    token,
+  };
+}
+
+/**
+ * Get the WebSocket URL for a DHE server URL.
+ * @param serverUrl The DHE server URL.
+ * @returns The WebSocket URL.
+ */
+export function getWsUrl(serverUrl: URL): URL {
+  const url = new URL('/socket', serverUrl);
+  if (url.protocol === 'http:') {
+    url.protocol = 'ws:';
+  } else {
+    url.protocol = 'wss:';
+  }
+  return url;
+}
+
+/**
+ * Determine if the logged in user has permission to interact with the UI.
+ * @param dheClient The DHE client.
+ * @returns A promise that resolves to true if the user has permission to interact with the UI.
+ */
+export async function hasInteractivePermission(
+  dheClient: EnterpriseClient
+): Promise<boolean> {
+  // TODO: Retrieve these group names from the server:
+  // https://deephaven.atlassian.net/browse/DH-9418
+  const GROUP_NON_INTERACTIVE = 'deephaven-noninteractive';
+  const GROUP_SUPERUSERS = 'iris-superusers';
+
+  const groups = await dheClient.getGroupsForUser();
+
+  const isSuperUser = groups.indexOf(GROUP_SUPERUSERS) >= 0;
+  const isNonInteractive = groups.indexOf(GROUP_NON_INTERACTIVE) >= 0;
+  const isInteractive = !isNonInteractive;
+
+  return isSuperUser || isInteractive;
+}
+
+/**
+ * Find the `QueryInfo` matching the given serial.
+ * @param dheClient DHE client to search.
+ * @param matchSerial Serial to match.
+ * @returns The matching `QueryInfo` or `undefined` if not found.
+ */
+export function findQueryConfigForSerial(
```

Comment:
`getWorkerInfoFromQuery` calls this function in response to `EVENT_CONFIG_UPDATED` events until it finds a match. Should that work, or does `EVENT_CONFIG_ADDED` need to be accounted for as well?

Reply:
Scanning all known configs on each update can be problematic with something like 10k+ queries since the update event is fired multiple times in the query lifecycle.
I think `EVENT_CONFIG_ADDED` needs to be handled instead of `EVENT_CONFIG_UPDATED`. Also check if event.details have the info you need so you don't have to re-scan the known configs.

Reply:
IIR, I had originally tried that one, but the worker url wasn't available when it was added. It only showed up after a future update. I could be misremembering though.

Reply:
Right, that seems possible. Was the worker URL eventually available in the update event details?

Reply:
yes